### DISCUSSION TITLE
Write text setting colors in render_mode=2

### DISCUSSION THREAD

**Discussion by pietro-foini** [pymupdf/PyMuPDF#2558] [Looking for help] [answered]
I am pasting text inside a PDF document using the `TextWriter` class, specifically using the following methods:

```
writer = fitz.TextWriter(page.rect) 
writer.fill_textbox(rect=rect, 
                    text=text, fontsize=font_size,
                    font=font,
                    pos=position,
                    align=0)
writer.write_text(page=page,
                  color=color,
                  render_mode=0)
```
Now, using the same methods, I would like to paste stroked text and be able to set both the stroke color and the fill color of the text. For example, having white stroke text and black fill text. Is it possible?

Looking at the documentation of `write_text` (https://pymupdf.readthedocs.io/en/latest/textwriter.html#TextWriter.color), it seems that setting `render_mode = 2` would be suitable for my case, but I haven't been able to set the two colors differently.

Similar example output text:

![image](https://github.com/pymupdf/PyMuPDF/assets/54908205/6b1d8fd0-796d-48bf-84fd-baf9778babba)

**Comment by JorjMcKie** [comment, marked as answer]
You can do this:
```python
import fitz
doc=fitz.open()
page=doc.new_page()
tw=fitz.TextWriter(page.rect)
text = "Hello world!"
rect = fitz.Rect(100,100,300,300)
tw1 = fitz.TextWriter(page.rect)
tw1.fill_textbox(rect,text)
tw1.write_text(page,color=(1,0,0),render_mode=0)
tw1.write_text(page,color=(0,0,1),render_mode=1)
doc.ez_save("x.pdf")
```
You can write the same TextWriter multiple times - on the same or different pages.
Looks like this:
![grafik](https://github.com/pymupdf/PyMuPDF/assets/8290722/cb748912-2fdb-4dbf-a6fa-6d1583ba1574)

**Reply by pietro-foini** [reply]
Thank you very much! The solution was indeed intuitive 😅. I take this opportunity to ask if, using the same methods, it is possible to manage the thickness of the text stroke.

**Reply by JorjMcKie** [reply]
> I take this opportunity to ask if, using the same methods, it is possible to manage the thickness of the text stroke.

No (not yet), sorry. For render mode != 0, I am setting the stroke line width to 5% of the font size: fontsize * 0.05.
But, since you are asking, it may make sense to support that as some parameter `linewidth` in `TextWriter.write_text()`.
